src/Decompiler/Loading/Loader.cs

```diff
@@ -83,47 +86,83 @@ public Program AssembleExecutable(string fileName, byte[] image, IAssembler asm,
+            if (project is not null)
+            {
+                project.FireScriptEvent(ScriptEvent.OnProgramLoaded);
+                return project;
+            }
+
+            var binaryImage = LoadBinaryImage(imageLocation, image, loaderName, addrLoad);
+            if (binaryImage is Program program)
+            {
+                project = new Project();
+                project.AddProgram(imageLocation, program);
+                project.LoadedMetadata = program.Platform.CreateMetadata();
+                program.EnvironmentMetadata = project.LoadedMetadata;
+                project.FireScriptEvent(ScriptEvent.OnProgramLoaded);
+                return project;
+            }
+            return binaryImage;
+        }
+
+        /// <summary>
+        /// Loads a binary image into memory, if it has a recognized file format.
+        /// </summary>
+        /// <param name="imageLocation">The location of the loaded image.</param>
+        /// <param name="image">The raw bytes fetched from <paramref name="imageLocation"/>.</param>
         /// <param name="loader">.NET Class name of a custom loader (may be null)</param>
         /// <param name="addrLoad">Address into which to load the file.</param>
-        /// <returns>A <see cref="ILoadedImage"/> if the file format is recognized, or null 
-        /// if the file cannot be recognized. Callers will have to handle the different
-        /// possible implementations of the interface.</returns>
-        public ILoadedImage? LoadImage(string filename, byte[] image, string? loader, Address? addrLoad)
+        /// <returns>A <see cref="ILoadedImage"/> if the file format is recognized, or 
+        /// an instance of <see cref="Blob"/> if the file cannot be recognized.</returns>
+        public ILoadedImage LoadBinaryImage(ImageLocation imageLocation, byte[] image, string? loader, Address? addrLoad)
         {
             ImageLoader? imgLoader;
             if (!string.IsNullOrEmpty(loader))
             {
-                imgLoader = CreateCustomImageLoader(Services, loader, filename, image);
+                imgLoader = CreateCustomImageLoader(Services, loader, imageLocation, image);
             }
             else
             {
-                imgLoader = FindImageLoader<ImageLoader>(filename, image);
-                if (imgLoader == null)
+                imgLoader = FindImageLoader<ImageLoader>(imageLocation, image);
+                if (imgLoader is null)
                 {
-                    imgLoader = CreateDefaultImageLoader(filename, image);
+                    imgLoader = CreateDefaultImageLoader(imageLocation, image);
                 }
             }
-            if (imgLoader == null)
-                return null;
+            if (imgLoader is null)
+                return new Blob(image);
 
-            if (addrLoad is null)
+            if (addrLoad is null && imgLoader is ProgramImageLoader ploader)
```

Comment:
Consider to refactor it to
```
if (imgLoader is ProgramImageLoader ploader)
{
    if (addrLoad is null)
    {
        addrLoad = ploader.PreferredBaseAddress;     //$REVIEW: Should be a configuration property.
    }
    var program = ploader.LoadProgram(addrLoad);
    return PostProcessProgram(imageLocation, addrLoad, ploader, program);
}
return imgLoader.Load(addrLoad);
```

Reply:
Thought it will be incorrect if `imgLoader` can return `Program`. Can it do it?

Reply:
This gets a bit confusing. A `ProgramImageLoader` must always return a `Program`. An `ImageLoader` _may_ return a `Program`, but it might also return a `Blob`, or an `IArchive`. We want to handle the case when a non-`ProgramImageLoader` returns a `Program` too.

Reply:
Will `PostProcessProgram` work correctly in that case? It will ignore relocations for program returned from non-`ProgramImageLoader`.